### ISSUE TITLE
Error 200: Special character used in LDAP query.

### ISSUE THREAD

**Issue by jay-hankins** [authelia/authelia#322]
I'm experiencing an interesting error. For some reason, the input sanitizer is thinking that my supplied username contains invalid/unsafe LDAP characters.

Here's a log from a request. Note that there is a log message that says the user was retrieved successfully. 
```
authelia_1        | 2019-02-07T05:24:03.648589158Z debug: date='Thu Feb 07 2019 05:24:03 GMT+0000 (UTC)' method='GET', path='/' requestId='9b3d225c-e0bb-4c4a-a905-f5975e970204' sessionId='nJHyFbsGe_yxdzXRIgjEvgXSXCdSHCpL' ip='10.42.69.1' message='Headers = {"host":"login.nutt.men","user-agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:66.0) Gecko/20100101 Firefox/66.0","accept":"text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8","accept-encoding":"gzip, deflate, br","accept-language":"en-US,en;q=0.5","cookie":"authelia_session=s%3AnJHyFbsGe_yxdzXRIgjEvgXSXCdSHCpL.n6Q63gSOc%2F3YuPmH6hU3G844FlvKTulgKuXp16e62Y0","dnt":"1","te":"trailers","upgrade-insecure-requests":"1","x-forwarded-for":"10.42.69.1","x-forwarded-host":"login.nutt.men","x-forwarded-port":"443","x-forwarded-proto":"https","x-forwarded-server":"31d9ecb1d9c1","x-real-ip":"10.42.69.1"}'
authelia_1        | 2019-02-07T05:24:07.492098307Z debug: date='Thu Feb 07 2019 05:24:07 GMT+0000 (UTC)' method='POST', path='/api/firstfactor' requestId='0f11b8bf-5c41-4840-9a39-af10527a904f' sessionId='nJHyFbsGe_yxdzXRIgjEvgXSXCdSHCpL' ip='10.42.69.1' message='Headers = {"host":"login.nutt.men","user-agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:66.0) Gecko/20100101 Firefox/66.0","content-length":"33","accept":"*/*","accept-encoding":"gzip, deflate, br","accept-language":"en-US,en;q=0.5","content-type":"application/x-www-form-urlencoded; charset=UTF-8","cookie":"authelia_session=s%3AnJHyFbsGe_yxdzXRIgjEvgXSXCdSHCpL.n6Q63gSOc%2F3YuPmH6hU3G844FlvKTulgKuXp16e62Y0","dnt":"1","referer":"https://login.nutt.men/","te":"trailers","x-forwarded-for":"10.42.69.1","x-forwarded-host":"login.nutt.men","x-forwarded-port":"443","x-forwarded-proto":"https","x-forwarded-server":"31d9ecb1d9c1","x-real-ip":"10.42.69.1","x-requested-with":"XMLHttpRequest"}'
authelia_1        | 2019-02-07T05:24:07.493670042Z info: date='Thu Feb 07 2019 05:24:07 GMT+0000 (UTC)' method='POST', path='/api/firstfactor' requestId='0f11b8bf-5c41-4840-9a39-af10527a904f' sessionId='nJHyFbsGe_yxdzXRIgjEvgXSXCdSHCpL' ip='10.42.69.1' message='Starting authentication of user "jhankins"'
authelia_1        | 2019-02-07T05:24:07.504563004Z info: date='Thu Feb 07 2019 05:24:07 GMT+0000 (UTC)' method='POST', path='/api/firstfactor' requestId='0f11b8bf-5c41-4840-9a39-af10527a904f' sessionId='nJHyFbsGe_yxdzXRIgjEvgXSXCdSHCpL' ip='10.42.69.1' message='No regulation applied.'
authelia_1        | 2019-02-07T05:24:07.517487186Z debug: LDAP: Bind user 'cn=admin,dc=nutt,dc=men'
authelia_1        | 2019-02-07T05:24:07.610301741Z debug: Computed users filter is cn=jhankins
authelia_1        | 2019-02-07T05:24:07.610508214Z debug: LDAP: searching for user dn of jhankins
authelia_1        | 2019-02-07T05:24:07.614189533Z debug: LDAP: retrieved user dn is cn=jhankins,ou=users,dc=nutt,dc=men
authelia_1        | 2019-02-07T05:24:07.624145170Z debug: LDAP: Bind user 'cn=jhankins,ou=users,dc=nutt,dc=men'
authelia_1        | 2019-02-07T05:24:07.719635120Z debug: LDAP: Unbind user 'cn=jhankins,ou=users,dc=nutt,dc=men'
authelia_1        | 2019-02-07T05:24:07.724223637Z debug: Computed users filter is cn=jhankins
authelia_1        | 2019-02-07T05:24:07.724591746Z debug: LDAP: searching for user dn of jhankins
authelia_1        | 2019-02-07T05:24:07.728201636Z debug: LDAP: Unbind user 'cn=admin,dc=nutt,dc=men'
authelia_1        | 2019-02-07T05:24:07.729643029Z debug: LDAP: retrieved user dn is cn=jhankins,ou=users,dc=nutt,dc=men
authelia_1        | 2019-02-07T05:24:07.735865454Z error: date='Thu Feb 07 2019 05:24:07 GMT+0000 (UTC)' method='POST', path='/api/firstfactor' requestId='0f11b8bf-5c41-4840-9a39-af10527a904f' sessionId='nJHyFbsGe_yxdzXRIgjEvgXSXCdSHCpL' ip='10.42.69.1' message='Reply with error 200: Special character used in LDAP query.'
authelia_1        | 2019-02-07T05:24:07.736356936Z debug: date='Thu Feb 07 2019 05:24:07 GMT+0000 (UTC)' method='POST', path='/api/firstfactor' requestId='0f11b8bf-5c41-4840-9a39-af10527a904f' sessionId='nJHyFbsGe_yxdzXRIgjEvgXSXCdSHCpL' ip='10.42.69.1' message='LdapError: Special character used in LDAP query.
authelia_1        | 2019-02-07T05:24:07.736515269Z     at that.withSession.catch (/usr/src/server/src/lib/authentication/backends/ldap/LdapUsersDatabase.js:32:45)
authelia_1        | 2019-02-07T05:24:07.736623166Z     at tryCatcher (/usr/src/node_modules/bluebird/js/release/util.js:16:23)
authelia_1        | 2019-02-07T05:24:07.736657752Z     at Promise._settlePromiseFromHandler (/usr/src/node_modules/bluebird/js/release/promise.js:512:31)
authelia_1        | 2019-02-07T05:24:07.736685546Z     at Promise._settlePromise (/usr/src/node_modules/bluebird/js/release/promise.js:569:18)
authelia_1        | 2019-02-07T05:24:07.736751118Z     at Promise._settlePromise0 (/usr/src/node_modules/bluebird/js/release/promise.js:614:10)
authelia_1        | 2019-02-07T05:24:07.736841118Z     at Promise._settlePromises (/usr/src/node_modules/bluebird/js/release/promise.js:690:18)
authelia_1        | 2019-02-07T05:24:07.736873933Z     at _drainQueueStep (/usr/src/node_modules/bluebird/js/release/async.js:138:12)
authelia_1        | 2019-02-07T05:24:07.736899654Z     at _drainQueue (/usr/src/node_modules/bluebird/js/release/async.js:131:9)
authelia_1        | 2019-02-07T05:24:07.736924405Z     at Async._drainQueues (/usr/src/node_modules/bluebird/js/release/async.js:147:5)
authelia_1        | 2019-02-07T05:24:07.736949129Z     at Immediate.Async.drainQueues (/usr/src/node_modules/bluebird/js/release/async.js:17:14)
authelia_1        | 2019-02-07T05:24:07.736972498Z     at runCallback (timers.js:785:20)
authelia_1        | 2019-02-07T05:24:07.736995229Z     at tryOnImmediate (timers.js:747:5)
authelia_1        | 2019-02-07T05:24:07.737018805Z     at processImmediate [as _immediateCallback] (timers.js:718:5)'
```

I'm unsure if I've misconfigured the LDAP connection, but it looks normal to me:
```
authentication_backend:
  # LDAP backend configuration.
  #
  # This backend allows Authelia to be scaled to more
  # than one instance and therefore is recommended for
  # production.
  ldap:
    # The url of the ldap server
    url: ldap://openldap

    # The base dn for every entries
    base_dn: dc=nutt,dc=men

    # An additional dn to define the scope to all users
    additional_users_dn: ou=users

    # The users filter used to find the user DN
    # {0} is a matcher replaced by username.
    # 'cn={0}' by default.
    users_filter: cn={0}

    # An additional dn to define the scope of groups
    # additional_groups_dn: ou=groups

    # The groups filter used for retrieving groups of a given user.
    # {0} is a matcher replaced by username.
    # {dn} is a matcher replaced by user DN.
    # 'member={dn}' by default.
    # groups_filter: (&(member={dn})(objectclass=groupOfNames))

    # The attribute holding the name of the group
    # group_name_attribute: cn

    # The attribute holding the mail address of the user
    mail_attribute: mail

    # The username and password of the admin user.
    user: cn=admin,dc=nutt,dc=men
    password: admin
```

Any ideas? I've traced that forbidden chars to here https://github.com/clems4ever/authelia/blob/2e087f12f40c01958144f8cba5f17ef69d5bbde5/server/src/lib/ldap/InputsSanitizer.ts#L16 but my username doesn't have any of those characters. What am I missing?

**Comment by clems4ever** [comment]
Hello @jay-hankins , this is indeed very weird... Did you create a ldap DB from a ldif file or do you already have a populated LDAP? In the first case please share the file (remove the password hashes).
Do you run the code from master? I can probably add a log to highlight the erroneous input if it can help.

**Comment by jay-hankins** [comment]
Hi @clems4ever, my particular approach with LDAP should be fairly standard. I run an openldap container, and a phpldapadmin container. So I created a user for myself from phpldapadmin.

Here are the relevant entries from `docker-compose.yml`, standard stuff:
```
  openldap-admin:
    image: osixia/phpldapadmin:0.6.11
    ports:
      - 9090:80
    environment:
      - PHPLDAPADMIN_LDAP_HOSTS=openldap
      - PHPLDAPADMIN_HTTPS=false
    networks:
      - auth-network
  
  openldap:
    image: osixia/openldap:1.2.1 
    environment:
      - LDAP_ORGANISATION=Nutt Ltd.
      - LDAP_DOMAIN=nutt.men
    networks:
      - auth-network
    hostname: openldap
```

Here's the LDIF exported from phpldapadmin:
```

version: 1

# Entry 1: dc=nutt,dc=men
dn: dc=nutt,dc=men
dc: nutt
o: Nutt Ltd.
objectclass: top
objectclass: dcObject
objectclass: organization

# Entry 2: cn=admin,dc=nutt,dc=men
dn: cn=admin,dc=nutt,dc=men
cn: admin
description: LDAP administrator
objectclass: simpleSecurityObject
objectclass: organizationalRole
userpassword: {SSHA}EWx+xRMkne0J9sfzOoXzZZYWxZEnGKbw

# Entry 3: ou=groups,dc=nutt,dc=men
dn: ou=groups,dc=nutt,dc=men
objectclass: organizationalUnit
objectclass: top
ou: groups

# Entry 4: cn=admin-users,ou=groups,dc=nutt,dc=men
dn: cn=admin-users,ou=groups,dc=nutt,dc=men
cn: admin-users
gidnumber: 501
objectclass: posixGroup
objectclass: top

# Entry 5: cn=media-users,ou=groups,dc=nutt,dc=men
dn: cn=media-users,ou=groups,dc=nutt,dc=men
cn: media-users
gidnumber: 500
objectclass: posixGroup
objectclass: top

# Entry 6: ou=users,dc=nutt,dc=men
dn: ou=users,dc=nutt,dc=men
objectclass: organizationalUnit
objectclass: top
ou: users

# Entry 7: cn=jhankins,ou=users,dc=nutt,dc=men
dn: cn=jhankins,ou=users,dc=nutt,dc=men
cn: jhankins
gidnumber: 501
givenname: Jay
homedirectory: /home/users/jhankins
objectclass: inetOrgPerson
objectclass: posixAccount
objectclass: top
sn: Hankins
uid: jhankins
uidnumber: 1000
userpassword: {SSHA}/
```

**Comment by clems4ever** [comment]
This will help troubleshoot further: https://github.com/clems4ever/authelia/pull/323

**Comment by jay-hankins** [comment]
Thanks for that PR. Here's what we've got:
```
error: Error with input jhankins. Cause:TypeError: Cannot read property 'indexOf' of undefined
```

Probably coming from here https://github.com/clems4ever/authelia/blob/23e28ee659cda65bc63fec5fc490e7b2da2af2cf/server/src/lib/authentication/backends/ldap/Sanitizer.ts#L5 but I'm not sure why the parameter isn't making through to the `contains` function.

**Comment by clems4ever** [comment]
@jay-hankins , sorry for the late reply but I did not manage to reproduce this issue. Can you send me more lines of logs so that I can see if something is wrong before the actual exception occurs?

**Comment by clems4ever** [comment]
@jay-hankins , any news?

**Comment by jay-hankins** [comment]
Hey @clems4ever, sorry for the delay. 

I believe this is all the relevant logging I can capture: 
https://gist.github.com/jay-hankins/f2aa7211f418d8a7c422325fbcc0f893

**Comment by seerickcode** [comment]
FYI, I had the same issue.  I also had group handling commented out.  As soon as I put proper group handling options back in, it worked.

**Comment by jay-hankins** [comment]
@seerickcode Boom, that was it.

**Comment by clems4ever** [comment]
@seerickcode , @jay-hankins , what options are you talking about exactly?

**Comment by jay-hankins** [comment]
@clems4ever  I had to uncomment `groups_filter` and `group_name__attribute`

**Comment by clems4ever** [comment]
Alright, it's maybe worth having a look and fix it or at least have an error message or a warning. Thanks for the report.

**Comment by jay-hankins** [comment]
No problem! Thanks for the friendly help and for the Authelia project 🙂

**Comment by nightah** [comment]
Fixed by 09b4e4e57e486a17dfd1371dc6c294e095dbdfb6.
